### ISSUE TITLE
StageConverter: Time looks like Zulu-Time, but isn't

### ISSUE THREAD

**Issue by ralfhergert** [gocd/gocd#3989]
https://github.com/gocd/gocd/blob/3cfb0a62fccabcc9e345df39cbede86c6cbea6db/plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/notification/v1/StageConverter.java#L31

The suffix 'Z' is defined as 'Zulu time' which means the time should be rendered as being in the timezone UTC+-0. But this pattern and how it is used renders in fact the time using the default locale. If the default locale is not the same as UTC+-0 appending the letter 'Z' creates in fact a wrong date.

A possible solution can be to actually include the correct timezone. For instance with the pattern:
`new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ")`
in which Z stands for an RFC822 timezone definition.

**Comment by arvindsv** [comment]
```
$ TZ=Europe/Paris jrunscript -l js -e "print((new java.text.SimpleDateFormat(\"yyyy-MM-dd'T'HH:mm:ss.SSSZ\")).format(new java.sql.Timestamp(java.lang.System.currentTimeMillis())))"
2017-11-07T17:43:56.852+0100
$ TZ=Europe/Paris jrunscript -l js -e "print((new java.text.SimpleDateFormat(\"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'\")).format(new java.sql.Timestamp(java.lang.System.currentTimeMillis())))"
2017-11-07T17:44:07.795Z

$ TZ=UTC jrunscript -l js -e "print((new java.text.SimpleDateFormat(\"yyyy-MM-dd'T'HH:mm:ss.SSSZ\")).format(new java.sql.Timestamp(java.lang.System.currentTimeMillis())))"
2017-11-07T16:44:19.677+0000
$ TZ=UTC jrunscript -l js -e "print((new java.text.SimpleDateFormat(\"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'\")).format(new java.sql.Timestamp(java.lang.System.currentTimeMillis())))"
2017-11-07T16:44:24.176Z
```

Agree.

/cc @bdpiparva Thoughts?

**Comment by bdpiparva** [comment]
@arvindsv, @ralfhergert ---  `yyyy-MM-dd'T'HH:mm:ss.SSS'Z'` always formats the timestamp in `UTC` and that is the expected behaviour. I would say `yyyy-MM-dd'T'HH:mm:ss.SSS'Z'` or `yyyy-MM-dd'T'HH:mm:ss.SSSZ` both points to same `System.currentTimeMillis()` what I mean is

```
$ TZ=Europe/Paris jrunscript -l js -e "print(new java.text.SimpleDateFormat(\"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'\").parse(\"2017-11-07T17:44:07.852Z\").getTime())"
1510073047852

$ TZ=Europe/Paris jrunscript -l js -e "print(new java.text.SimpleDateFormat(\"yyyy-MM-dd'T'HH:mm:ss.SSSZ\").parse(\"2017-11-07T17:44:07.852+0100\").getTime())"
1510073047852
```

Let's see how javascript reacts to this time formats

```
$ new Date('2017-11-07T17:44:07.852Z')
Tue Nov 07 2017 23:14:07 GMT+0530 (IST)
$ new Date('2017-11-07T17:44:07.852+0100')
Tue Nov 07 2017 22:14:07 GMT+0530 (IST)
```

I think we should not change the current implementation. @arvindsv --- WDYT ?

**Comment by arvindsv** [comment]
@bdpiparva I was more worried about the fact that the actual stage status change time is the "local time" stored in the DB (see [this](https://github.com/bdpiparva/gocd/blob/e54970569a0b7b158adda015fb017d73c8e47707/plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/notification/v1/StageNotificationDTO.java#L140), [this](https://github.com/bdpiparva/gocd/blob/e54970569a0b7b158adda015fb017d73c8e47707/plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/notification/v1/StageConverter.java#L57) and [this](https://github.com/bdpiparva/gocd/blob/e54970569a0b7b158adda015fb017d73c8e47707/domain/src/com/thoughtworks/go/domain/Stage.java#L69), which leads to [this](https://github.com/bdpiparva/gocd/blob/e54970569a0b7b158adda015fb017d73c8e47707/base/src/com/thoughtworks/go/util/SystemTimeClock.java#L45)).

So, at time 10:00AM IST (ie, +0530), if a stage finishes, won't a notification be sent saying that the notification happened at 10:00AM UTC? Since the 'Z' is just added at the end, and no conversion of that time to UTC really happens.

**Comment by ralfhergert** [comment]
@bdpiparva Yes, you are correct. If the timezone on the SimpleDateFormat ist set explicitly to UTC, then it is valid to append the letter 'Z' at the end of the rendered date to mark it as a Zulu time.
```
SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
format.setTimeZone(TimeZone.getTimeZone("UTC"));
```
